nomad/job_endpoint.go

```diff
@@ -328,52 +349,51 @@ func (j *Job) Register(args *structs.JobRegisterRequest, reply *structs.JobRegis
 			return err
 		}
 
+		submittedEval = true
+
 		// Populate the reply with job information
 		reply.JobModifyIndex = index
+		reply.Index = index
+
+		if args.Eval != nil {
+			reply.EvalCreateIndex = index
+		}
+
 	} else {
 		reply.JobModifyIndex = existingJob.JobModifyIndex
 	}
 
 	// used for multiregion start
 	args.Job.JobModifyIndex = reply.JobModifyIndex
 
-	// If the job is periodic or parameterized, we don't create an eval.
-	if args.Job.IsPeriodic() || args.Job.IsParameterized() {
+	if args.Eval == nil {
 		return nil
 	}
 
-	// Create a new evaluation
-	now := time.Now().UTC().UnixNano()
-	eval := &structs.Evaluation{
-		ID:             uuid.Generate(),
-		Namespace:      args.RequestNamespace(),
-		Priority:       args.Job.Priority,
-		Type:           args.Job.Type,
-		TriggeredBy:    structs.EvalTriggerJobRegister,
-		JobID:          args.Job.ID,
-		JobModifyIndex: reply.JobModifyIndex,
-		Status:         structs.EvalStatusPending,
-		CreateTime:     now,
-		ModifyTime:     now,
-	}
-	update := &structs.EvalUpdateRequest{
-		Evals:        []*structs.Evaluation{eval},
-		WriteRequest: structs.WriteRequest{Region: args.Region},
-	}
+	// COMPAT(1.1.0): Remove the ServerMeetMinimumVersion check.
+	// 0.12.1 introduced atomic eval job registration
+	if args.Eval != nil &&
+		!(submittedEval && ServersMeetMinimumVersion(j.srv.Members(), minJobRegisterAtomicEvalVersion, false)) {
```

Comment:
If we did the ServersMeetMinimumVersion check before submitting an atomic job+eval could we skip all of the dupe-eval checking code? Seems like that might be simpler than filtering dupes.

Reply:
Checking ServersMeetMinimumVersion is not sufficient.  Consider 3 clusters with A (leader) and B running 0.12.1, and C runs 0.12.0: the leader A will need to emit the eval update for C to handle, but B needs to de-dup it.  Once all servers upgrade to 0.12.1, they must dedup whenever they re-process the log (due to server restarts) until all these log entries get snapshotted.  This is another manifestation of https://github.com/hashicorp/nomad/issues/6934 .

Reply:
Hm, I don't see why that's the case if the logic is:

```
if ServersMeetMinimumVersion("0.12.1") {
  atomicJobAndEval()
} else {
  updateJob()
  updateEval()
}
```

ServersMeetMinimumVersion ensures *all* servers meet a minimum version, so until all servers are >=0.12.1 you would continue emitting the old form.

Once all servers are upgraded you would only emit the new form.

Since the check is done once-per-job-update, each job+eval will only ever be submitted atomically or individually -- never both.

Reply:
Good point.  I ruled out this approach early on because I was eager to get as close to atomic behavior as soon as 0.12.1 becomes a leader - but seeing how the complications, it does make sense to switch after all servers upgrade.  Change incoming.

Reply:
Done!  I like how it turned out a bit more!  Thank you for the suggestion!